In [2]:
import numpy as np
import pandas as pd

import cvxpy as cp

# Task 1
$$p^* = \max Z = 5x_1 + 4x_2$$
$$\begin{cases} 3x_1 + 4x_2 \leq 10 \\ x_1, x_2 \in \mathbb{N} \geq 0 \end{cases}$$

In [15]:
x_1 = cp.Variable()
x_2 = cp.Variable()

obj = cp.Maximize(5*x_1 + 4*x_2)
prob = cp.Problem(obj,
                 [
                     3*x_1 + 4*x_2 <= 10,
                     x_1 >= 0,
                     x_2 >= 0
                 ])
prob.solve()

print("\nThe optimal value is", prob.value)
print(f"x1: {x_1.value}, x2: {x_2.value}")


The optimal value is 16.66666664955361
x1: 3.333333326749174, x2: 3.951934089095033e-09


### Node 1
UB: $16.6666$

LB: $15\ (x_1 = 3, x_2 = 0)$

### Node 1.1
$x_1 \leq 3$

### Node 1.2
$x_1 \geq 4$ is not possible because first constraint is not met

In [16]:
prob = cp.Problem(obj,
                 [
                     3*x_1 + 4*x_2 <= 10,
                     x_1 <= 3,
                     x_1 >= 0,
                     x_2 >= 0
                 ])
prob.solve()

print("\nThe optimal value is", prob.value)
print(f"x1: {x_1.value}, x2: {x_2.value}")


The optimal value is 15.99999999955193
x1: 2.999999999934109, x2: 0.24999999997034605


### Node 1.1
UB: $16$

LB: $15\ (x_1 = 3, x_2 = 0)$

### Node 1.1.1
$x_2 >= 1$

### Node 1.1.2
$x_2 <= 0$

In [20]:
# 1.1.1
prob = cp.Problem(obj,
                 [
                     3*x_1 + 4*x_2 <= 10,
                     x_1 <= 3,
                     x_1 >= 0,
                     x_2 >= 1
                 ])
prob.solve()

print("1.1.1:")
print("\nThe optimal value is", prob.value)
print(f"x1: {x_1.value}, x2: {x_2.value}")

# 1.1.2
prob = cp.Problem(obj,
                 [
                     3*x_1 + 4*x_2 <= 10,
                     x_1 <= 3,
                     x_1 >= 0,
                     x_2 >= 0,
                     x_2 <= 0
                 ])
prob.solve()

print("1.1.2:")
print("\nThe optimal value is", prob.value)
print(f"x1: {x_1.value}, x2: {x_2.value}")

1.1.1:

The optimal value is 13.999999989067822
x1: 1.9999999961387218, x2: 1.0000000020935529
1.1.2:

The optimal value is 15.000000000565414
x1: 3.000000000111104, x2: 2.4729501896409894e-12


### Node 1.1.1
UB: $14$

LB: $14$

### Node 1.1.2

UB: $15$

LB: $15$

Optimal solution is found at node 1.1.2 because the upper and lower bounds match and the lower bound is the highest value of all nodes. Thus, the solution is $x_1 = 3, x_2 = 0$.

# Task 2
$$p^* = \max \text{profit} = 50*\text{coat} + 40*\text{slack}$$
$$\begin{cases} 3*\text{coat} + 5*\text{slack} \leq 150 \\ 10*\text{coat} + 4*\text{slack} \leq 200 \end{cases}$$

In [25]:
coat = cp.Variable()
slack = cp.Variable()

profit = cp.Maximize(50*coat + 40*slack)
prob = cp.Problem(profit,
                 [
                     3*coat + 5*slack <= 150,
                     10*coat + 4*slack <= 200,
                     coat >= 0,
                     slack >= 0,
                 ])
prob.solve()

print("\nThe highest profit is:", prob.value)
print(f"n_coat: {coat.value}, n_slack: {slack.value}")


The highest profit is: 1473.6842103374033
n_coat: 10.52631578438763, n_slack: 23.684210527950544


### Node 1
UB: $1473.684$

LB: $1420\ (50*10 + 40*23)$

Take the node with largest deviation from its floor value: 23.68
### Node 1.1
$\text{slack} \leq 23$

### Node 1.2
$\text{slack} \geq 24$

In [30]:
# 1.1
prob = cp.Problem(profit,
                 [
                     3*coat + 5*slack <= 150,
                     10*coat + 4*slack <= 200,
                     coat >= 0,
                     slack >= 0,
                     slack <= 23
                 ])
prob.solve()

print("1.1:")
print("The highest profit is:", prob.value)
print(f"n_coat: {coat.value}, n_slack: {slack.value}")

# 1.2
prob = cp.Problem(profit,
                 [
                     3*coat + 5*slack <= 150,
                     10*coat + 4*slack <= 200,
                     coat >= 0,
                     slack >= 24
                 ])
prob.solve()

print("\n1.2:")
print("The highest profit is:", prob.value)
print(f"n_coat: {coat.value}, n_slack: {slack.value}")

1.1:
The highest profit is: 1459.9999998370172
n_coat: 10.800000000865548, n_slack: 22.99999999484349

1.2:
The highest profit is: 1460.0000000417604
n_coat: 9.999999997420302, n_slack: 24.000000004268625


### Node 1.1
UB: $1460$

LB: $1380\ (50*10 + 40*22)$

This node's lower bound dropped below the original lower bound so we're not interested in it.

### Node 1.2
UB: $1460$

LB: $1460\ (50*10 + 40*24)$

We converged on a solution here because UB = LB, and it's maximal LB.

Then tailor should produce 10 coats and 24 slacks.